tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/SetMembershipSimplifier.scala

```diff
@@ -0,0 +1,51 @@
+package at.forsyte.apalache.tla.pp
+
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.lir.oper._
+import at.forsyte.apalache.tla.lir.transformations.standard.FlatLanguagePred
+import at.forsyte.apalache.tla.lir.transformations.{LanguageWatchdog, TransformationTracker}
+import at.forsyte.apalache.tla.lir.values._
+
+/**
+ * A simplifier that rewrites vacuously true membership tests based on type information.
+ *
+ * For example, `x \in BOOLEAN` is rewritten to `TRUE` if x is typed BoolT1.
+ *
+ * @author
+ *   Thomas Pani
+ */
+class SetMembershipSimplifier(tracker: TransformationTracker) extends AbstractTransformer(tracker) {
+  private val boolTag = Typed(BoolT1())
+  private val trueVal = ValEx(TlaBool(true))(boolTag)
+
+  override val partialTransformers = List(transformMembership)
+
+  override def apply(expr: TlaEx): TlaEx = {
+    LanguageWatchdog(FlatLanguagePred()).check(expr)
+    transform(expr)
+  }
+
+  private def transformMembership: PartialFunction[TlaEx, TlaEx] = {
+    case OperEx(TlaSetOper.in, name, ValEx(TlaBoolSet)) if name.typeTag == Typed(BoolT1()) => trueVal
+    case OperEx(TlaSetOper.in, name, ValEx(TlaStrSet)) if name.typeTag == Typed(StrT1())   => trueVal
+    case OperEx(TlaSetOper.in, name, ValEx(TlaIntSet)) if name.typeTag == Typed(IntT1())   => trueVal
+    case OperEx(TlaSetOper.in, name, ValEx(TlaRealSet)) if name.typeTag == Typed(RealT1()) => trueVal
+    case OperEx(TlaSetOper.in, name, OperEx(TlaSetOper.seqSet, ValEx(TlaBoolSet)))
+        if name.typeTag == Typed(SeqT1(BoolT1())) =>
+      trueVal
+    case OperEx(TlaSetOper.in, name, OperEx(TlaSetOper.seqSet, ValEx(TlaStrSet)))
+        if name.typeTag == Typed(SeqT1(StrT1())) =>
+      trueVal
+    case OperEx(TlaSetOper.in, name, OperEx(TlaSetOper.seqSet, ValEx(TlaIntSet)))
+        if name.typeTag == Typed(SeqT1(IntT1())) =>
+      trueVal
+    case OperEx(TlaSetOper.in, name, OperEx(TlaSetOper.seqSet, ValEx(TlaRealSet)))
+        if name.typeTag == Typed(SeqT1(RealT1())) =>
+      trueVal
```

Comment:
I feel like there's some unnecessary repetition here. You could write something along the lines of just:
```scala
    case OperEx(TlaSetOper.in, name, set) if isMatch(name.typeTag.asTlaType1(), set) => trueEx
    case ex => ex
```
where
```scala
def typeOfElem(ps: TlaPredefSet): TlaType1 = ...
def isMatch(nameType, set) = set match {
  case ValEx(ps: TlaPredefSet) => nameType == typeOfElem(ps)
  case OperEx(TlaSetOper.seqSet, ValEx(ps: PredefSet)) => nameType == SeqT1(typeOfElem(ps))
  case _ => false 
}
```

Reply:
I prefer my solution; it's more verbose, but also more readable.

Reply:
I'm in favor of @Kukovec's suggestion. I'm not worried about LOC here if it impacts readability, but having to manually repeat the same logic `n` times introduces `n` chances for errors to creep in. @Kukovec's suggestion is also much better for maintainability, as it makes it possible to add new operators under `in`, or change the names of operators or types, without needing to update the case analysis here. 

A compromise might be something like:

```suggestion
    case OperEx(TlaSetOper.in, name, ValEx(valEx)) if name.typeTag.asTlaType1() == typeOfElem(valEx) => trueVal
    case OperEx(TlaSetOper.in, name, OperEx(TlaSetOper.seqSet, ValEx(valEx))) if name.typeTag.asTlaType1 == SeqT1(typeOfElem(valEx) =>
```

But either way, I think manually having to type out the correlation of terms with their types is illdvised when an automatic solution is readily available.

Reply:
@thpani @Kukovec You both assume that code is more often written/edited than read; in my experience this is almost never true.

But Jure's comment inspired me to do [aac5a6f](https://github.com/informalsystems/apalache/pull/1406/commits/aac5a6fd3d84b2bf2f56833045f3f09151a47e8c), which is similar to Shon's suggestion.

Reply:
I don't think our motives here are primarily about privileging writing over reading. I also don't agree, personally, that having 15 lines of the same logic, with small variations, is more readable than having two lines that clearly communicate the abstract pattern. But readability is a pretty subjective, and what is readable varies from person to person, and in the same person depending on their familiarity and experiences.

However, the difference between doing stuff manually and having clean, automated abstractions seems pretty objective I think. And, IOM, even if it were proven that extensional definitions of functions were more readable by some objective standard, I still think we'd have good reason to favor intensional definitions for any functions with more than 4 or 5 elements in the domain. That is, even if we could measure readability objectively, I think I'd still be in favor of clean abstractions over dozens of lines of repetition :)

Reply:
Thanks for making this change! I find it much easier to parse out at a glance.

Reply:
Look, it's about good abstractions, and I like the one we arrived at :)

All I was saying is that abstractions in computer programming aren't pure mathematical objects and come at a (in this case) cognitive cost. For me, there's definitely a trade-off between copy-pasting a line (that gets code review) versus spatially locating and loading/unloading an abstractions into my brain.